Add TLS/mTLS support for Tornjak (#338)
spiffe/helm-charts: charts/spire/charts/spire-server/README.md

* <https://github.com/spiffe/helm-charts/tree/main/charts/spire>

## Tornjak

> Tornjak is the UI and Control Plane for SPIRE [https://github.com/spiffe/tornjak](https://github.com/spiffe/tornjak)

Reviewer: Should we mention the UI part in this README? 

Technically it is part of another chart. Probably better to document this in the spire chart or the tornjak-ui chart. And refer int this documentation to that chart.

When Tornjak is enabled, it can be reached in one of the three connection types:

* http - simple HTTP connection recommended for demonstration purpose only. It should not be use for production.
* tls - enables Transport Layer Security (TLS) with HTTPS communication. Tornjak client can validate Tornjak server using TLS Certificate (see instructions below)
* mtls - mutual TLS. Both, Tornjak client and the server can validate each other using provided certificates (see instructions below)

### Tornjak with TLS

`tornjak.config.connectionType=tls`

TLS connection requires Tornjak to have access to TLS key and certificate.
Complete instruction on creating your own TLS certificate can be found [here](https://github.com/spiffe/tornjak/blob/main/examples/tls_mtls/README.md).

> TLS Certificate and the private key must be provided to Tornjak via *TLS Secret*. Prior to deploying this Helm chart, create TLS Secret in the deployment namespace (e.g. `spire-server`)

Reviewer: Shouldn't we just bake this in, similar like the service accounts are implemented?

* <https://github.com/spiffe/helm-charts/tree/main/charts/spire>

## Tornjak

Tornjak is the UI and Control Plane for SPIRE [https://github.com/spiffe/tornjak](https://github.com/spiffe/tornjak) and it is composed of two components:

* Backend (this chart) - Tornjak APIs that extend SPIRE APIs with Control Plane functionality
* [Frontend](../tornjak-frontend/README.md) - Tornjak UI

> When Tornjak is enabled, it can be reached in one of the three connection types:
> 
> * HTTP - simple HTTP connection recommended for demonstration purpose only. It should not be use for production.
> * TLS - enables Transport Layer Security (TLS) with HTTPS communication. Tornjak client can validate Tornjak server using TLS Certificate (see instructions below)
> * mTLS - mutual TLS. Both, Tornjak client and the server can validate each other using provided certificates (see instructions below)

Reviewer: Here a suggestion to update the docs in such way we should also implement this feature to improve the best UX (User Experience) for both the chart installation as well usage of Tornjak itself.

```suggestion
When Tornjak is enabled, it is exposed on both http and https (if TLS server certs are configured). Tornjak handles a permanent redirect from `http` to `https` to ensure users always use the https endpoint.

In addition to that you can configure a `client certificate authority`, this will make Tornjak backend verify Client certificates signed by this authority to enable mTLS authentication.

> **Warning**: For production ensure to configure the certificates including CA, to protect Tornjak from unauthorized access.
```

Reviewer 2: tls is highly recommended. mtls is more of an authentication feature. When oidc is implemented, that may be preferred to be used over mtls in a lot of cases, so we shouldn't necessarily strongly recommend the user ca part.

Reviewer: Agree, but for now it is the only option and therefore the current recommendation until we have OIDC available.

Reviewer 2: Yeah. I guess we can amend it once that's implemented to recommend specifying client CA, or OIDC at that point.

* Backend (this chart) - Tornjak APIs that extend SPIRE APIs with Control Plane functionality
* [Frontend](../tornjak-frontend/README.md) - Tornjak UI

When Tornjak is enabled, it is exposed on both http and https (if TLS server certs are configured). Tornjak handles a permanent redirect from `http` to `https` to ensure users always use the https endpoint.

In addition to that you can configure a `client certificate authority`, this will make Tornjak backend verify Client certificates signed by this authority to enable mTLS authentication.

> > **Warning**: For production ensure to configure the certificates including CA, to protect Tornjak from unauthorized access.

Reviewer 2: ```suggestion
> **Warning**: For production, we recommend configuring tls certificates and client CA to protect Tornjak from unauthorized access.
```

| tornjak.config.http.port | int | `10000` | Container port value for HTTP |
| tornjak.config.http.service | object | `{"annotations":{},"port":10000,"type":"ClusterIP"}` | Service to handle Tornjak HTTP connection |
| tornjak.config.mtls.ca | string | `"/opt/spire/sample-keys/rootCA.pem"` | CA for mTLS connection |
| tornjak.config.mtls.cert | string | `"/opt/spire/sample-keys/cert.pem"` | Certificate for mTLS connection |
| tornjak.config.mtls.enabled | bool | `false` | Enables Tornjak TLS service |
| tornjak.config.mtls.key | string | `"/opt/spire/sample-keys/key.pem"` | Key for mTLS connection |
| tornjak.config.mtls.port | int | `30000` | Container port value for mTLS |
| tornjak.config.mtls.service | object | `{"annotations":{},"port":30000,"type":"ClusterIP"}` | Service to handle Tornjak mTLS connection |
| tornjak.config.tls.cert | string | `"/opt/spire/sample-keys/cert.pem"` | Certificate for TLS connection |
| tornjak.config.tls.enabled | bool | `false` | Enables Tornjak TLS service |
| tornjak.config.tls.key | string | `"/opt/spire/sample-keys/key.pem"` | Key for TLS connection |
| tornjak.config.tls.port | int | `20000` | Container port value for TLS |

> | tornjak.config.tls.service | object | `{"annotations":{},"port":20000,"type":"ClusterIP"}` | Service to handle Tornjak TLS connection |

Reviewer 2: are http/mtls/tls exclusive or can you really enable more then one?

Author: You have to enable AT LEAST one, otherwise there is no point of running it, since you can't communicate, therefore if none is configured, Tornjak startup configuration validation throws an error

Reviewer 2: Is there reason to have more then one though? Just wondering if this should be a mode enum rather then a set of bools.

Reviewer 2: maybe an alternate angle... The mtls part, is that for user auth?

Author: As is, it can be used for user authentication, however I am preparing a next PR to bring a proper user authentication and authorization via Keycloak.

Author: BTW I liked your suggestion of separating the userCA into a second secret. It's done.
